File: Sources/Engine/editor/SKKOkuriEditor.cpp
/* -*- C++ -*-

  MacOS X implementation of the SKK input method.

  Copyright (C) 2008 Tomotaka SUWA <t.suwa@mac.com>

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

*/

#import <AquaSKKEngine/AquaSKKEngine-Preamble.h>
#import <AquaSKKEngine/SKKInputContext.h>
#import <AquaSKKEngine/SKKOkuriEditor.h>
#import <AquaSKKEngine/AquaSKKEngine-Swift.h>

SKKOkuriEditor::SKKOkuriEditor(SKKInputContext *context, SKKOkuriListener *listener)
    : SKKBaseEditor(context), impl_(new SwiftObject(AquaSKKEngine::SKKOkuriEditorImpl::init(context, listener))) {}

void SKKOkuriEditor::ReadContext() {
    (*impl_)->readContext();
}

void SKKOkuriEditor::WriteContext() {
    (*impl_)->writeContext();
}

void SKKOkuriEditor::Input(const std::string &fixed, const std::string &input, char code) {
    (*impl_)->input(fixed, input, code);
}

void SKKOkuriEditor::Input(SKKBaseEditorEvent event) {
    (*impl_)->bridgeInputEvent(event);
}

void SKKOkuriEditor::Commit(std::string &queue) {
    (*impl_)->commit(queue);
}

bool SKKOkuriEditor::IsOkuriComplete() const {
    return (*impl_)->isOkuriComplete();
}

File: Sources/Engine/editor/SKKOkuriEditor.swift
//
//  SKKOkuriEditor.swift
//  AquaSKKEngine
//
//  Created by mzp on 2025/03/05.
//

public class SKKOkuriEditorImpl: SKKEditorProtocol {
    let context: SKKInputContext
    let listener: SKKOkuriListenerProtocol
    var first: Bool
    var prefix: String
    var okuri: String
    var input: String

    public init(context: SKKInputContext, listener: SKKOkuriListenerProtocol) {
        self.context = context
        self.listener = listener
        first = false
        prefix = ""
        okuri = ""
        input = ""
    }

    public func readContext() {
        first = true
        prefix.removeAll()
        okuri.removeAll()
    }

    public func writeContext() {
        context.output.Compose(std.string("*\(okuri)"), 0)
        update()
    }

    public func input(ascii _: String) {}

    public func input(fixed: String, input: String, code: Int) {
        self.input = input
        if first {
            first = false
            if let newElement = Unicode.Scalar(UInt32(code)) {
                prefix += String(newElement)
            }

            // KesSi 対応
            if !fixed.isEmpty, !input.isEmpty {
                listener.okkuriListenerAppendEntry(fixed: fixed)
                update()
                return
            }
        }
        // fixed が ascii の場合には送りとはみなさない
        if !(fixed.first?.isASCII ?? true) {
            okuri += fixed
        }

        // OWsa 対応
        if okuri.isEmpty {
            prefix.removeAll()
            if input.isEmpty {
                if let newElement = Unicode.Scalar(UInt32(code)) {
                    prefix += String(newElement)
                }
            } else {
                if let newElement = input.lowercased().first {
                    prefix += String(newElement)
                }
            }
        } else {
            if prefix.isEmpty {
                if let newElement = Unicode.Scalar(UInt32(code)) {
                    prefix += String(newElement)
                }
            }
        }
        update()
    }

    public func bridgeInputEvent(_ rawValue: UInt32) {
        let event = SKKBaseEditorEvent(rawValue: rawValue)
        inputEvent(event: event)
    }

    public func inputEvent(event: SKKBaseEditorEvent) {
        if event == SKKBaseEditorEventBackSpace {
            if okuri.isEmpty {
                context.needs_setback = true
            } else {
                okuri.removeLast()
            }
        }
        update()
    }

    public func commit(queue: String) -> String {
        prefix.removeAll()
        okuri.removeAll()
        return queue
    }

    public func isOkuriComplete() -> Bool {
        return !okuri.isEmpty && input.isEmpty
    }

    private func update() {
        context.entry.SetOkuri(std.string(prefix), std.string(okuri))
    }
}
